### PR TITLE
crimson/osd: fix unexpected connection markdown in heartbeat

### PR DESCRIPTION
**Pull request by cyx1231st**
Pass reference when log Heartbeat::Connection instance, or the
destructor will be called incorrectly, and the conn be marked down
unexpectedly.

Also, the replacing conn is actually connected during replacement-reset
event.

Fixes: https://tracker.ceph.com/issues/47124

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
@xxhdx1985126 could you take a look?

**Review by tchaikov — APPROVED** [review]
these changes are neat. thanks!

**Comment by tchaikov** [comment]
will merge once @xxhdx1985126 approve this change.

**Comment by xxhdx1985126** [comment]
LGTM:-)

**Comment by cyx1231st** [comment]
Wait, something wrong in the unittest...

**Comment by tchaikov** [comment]
```
INFO  2020-08-27 04:09:43,002 [shard 0] ms - [Test] wait_ready(): wait for pending_send=0, pending_peer_receive=1, pending_receive=0, pending 1/0 ready/establish connections ...
INFO  2020-08-27 04:09:43,002 [shard 0] ms - [Test] connect_attempts=1
INFO  2020-08-27 04:09:43,002 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9012/2 >> osd.? v2:127.0.0.1:9014/4] intercepted SOCKET_CONNECTING(1)
INFO  2020-08-27 04:09:43,003 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9014/4 >> unknown.? -@58753] ProtocolV2::start_accept(): target_addr=127.0.0.1:58753/0
INFO  2020-08-27 04:09:43,004 [shard 0] ms - [0] [osd.2(Test) v2:127.0.0.1:9012/2@0 >> osd.? v2:127.0.0.1:9014/4] intercepted BANNER_WRITE(1) => FAULT
INFO  2020-08-27 04:09:43,004 [shard 0] ms - [Test] got FAULT
INFO  2020-08-27 04:09:44,823 [shard 0] ms - [osd.2(Test) v2:127.0.0.1:9012/2@0 >> osd.? v2:127.0.0.1:9014/4] execute_connecting(): fault at CONNECTING, going to WAIT -- std::system_error (error crimson::net:3, negotiation failure)
WARN  2020-08-27 04:09:45,157 [shard 0] ms - [osd.2(Test) v2:127.0.0.1:9012/2@0 >> osd.? v2:127.0.0.1:9014/4] waiting 0.2 seconds ...
INFO  2020-08-27 04:09:45,158 [shard 0] ms - [osd.2(Test) v2:127.0.0.1:9012/2@0 >> osd.? v2:127.0.0.1:9014/4] write_event: delay ...
INFO  2020-08-27 04:09:45,160 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9014/4 >> unknown.? -@58753] execute_accepting(): fault at ACCEPTING, going to CLOSING -- std::system_error (error crimson::net:4, read eof)
INFO  2020-08-27 04:09:45,160 [shard 0] ms - [osd.4(TestPeer) v2:127.0.0.1:9014/4 >> unknown.? -@58753] closing: reset no, replace no
INFO  2020-08-27 04:09:45,160 [shard 0] ms - [TestPeer] got reset from Test
ERROR 2020-08-27 04:09:46,543 [shard 0] none - /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/crimson/test_messenger.cc:1412 : In function 'virtual void {anonymous}::FailoverSuitePeer::ms_handle_reset(crimson::net::ConnectionRef, bool)', ceph_assert(%s)
tracked_conn == conn
Aborting on shard 0.
Backtrace:
  /usr/lib/x86_64-linux-gnu/libasan.so.5+0x000000000006bb2f
  0x000000000c7cb4f8
  0x000000000c7a3e67
  0x000000000c65aa7e
  0x000000000c65abf1
  0x000000000c6bcd85
  0x000000000c6f5406
  0x000000000c6f54a1
  /lib/x86_64-linux-gnu/libpthread.so.0+0x000000000001288f
  /lib/x86_64-linux-gnu/libc.so.6+0x000000000003ee96
  /lib/x86_64-linux-gnu/libc.so.6+0x0000000000040800
  0x000000000be9b880
  0x0000000009f6103b
  0x000000000a7317b9
  0x000000000a4aaf24
  0x000000000a5fb706
  0x000000000a6774cb
  0x000000000a66f6d8
  0x000000000a66fde7
  0x000000000a669db8
  0x000000000a65cc7b
  0x000000000a670061
  0x000000000a2f66b9
  0x000000000a2d90a0
  0x000000000a2f67b0
  0x000000000a2d8b77
  0x000000000a3787b5
  0x000000000c6a11c4
  0x000000000c6a5752
  0x000000000c6aac7c
  0x000000000c51e58e
  0x000000000c51c405
  0x000000000c51c9a5
  0x0000000009fcd3c9
  /lib/x86_64-linux-gnu/libc.so.6+0x0000000000021b96
  0x0000000009f27f39
```

	194 - unittest_seastar_messenger (Child aborted)

@cyx1231st  could you take a look?

**Comment by tchaikov** [comment]
> Wait, something wrong in the unittest...

ahh, exactly, just saw your comment.

File: src/crimson/net/Protocol.cc
Comment (cyx1231st) [inline src/crimson/net/Protocol.cc:75]: The check `if (dispatch_reset)` is missing.

Comment (cyx1231st) [reply src/crimson/net/Protocol.cc]: fixed.

**Pull request merged**
